### ISSUE TITLE
[java] CommentDefaultAccessModifier false positive on nested interface

### ISSUE THREAD

**Issue by harbulot** [pmd/pmd#4555]
**Affects PMD Version:** 7.0.0-rc2

**Rule:** CommentDefaultAccessModifier

https://pmd.github.io/pmd/pmd_rules_java_codestyle.html#commentdefaultaccessmodifier

**Description:**

* [`UnnecessaryModifier`](https://pmd.github.io/pmd/pmd_rules_java_codestyle.html#unnecessarymodifier) recommends to remove `public static` on a nested `interface` defined within a class.
* If we do that, [`CommentDefaultAccessModifier`](https://pmd.github.io/pmd/pmd_rules_java_codestyle.html#commentdefaultaccessmodifier), then flags that interface definition as missing a comment regarding the default access modifier.

**Code Sample demonstrating the issue:**

```java
public class Example {
    interface TestInterfaceA { // Flags CommentDefaultAccessModifier
        void noop();
    }
    
    static interface TestInterfaceB { // Flags CommentDefaultAccessModifier and UnnecessaryModifier
        void noop();
    }
    
    public static interface TestInterfaceC { // Flags UnnecessaryModifier
        void noop();
    }
}
```

**Expected outcome:**

A nested interface without any modifier (example `TestInterfaceA` above) should not trigger a `CommentDefaultAccessModifier` violation. (This is similar to issues #1211 and #1257 that were reported against version 6.5.0 and possibly fixed in version 6.6.0 with PR #1213.)

**Running PMD through:** *Maven* (using `maven-pmd-plugin` version *3.20.1-pmd-7-SNAPSHOT* configured with PMD *7.0.0-rc2*)

**Comment by oowekyala** [comment]
> [UnnecessaryModifier](https://pmd.github.io/pmd/pmd_rules_java_codestyle.html#unnecessarymodifier) recommends to remove public static on a nested interface defined within a class.

I think this is the real false positive though. The `public` is not unnecessary. If you don't remove the public then CommentDefaultAccessModifier should be quiet.

**Comment by harbulot** [comment]
@oowekyala Yes, thank you, you're right.

**Comment by jsotuyod** [comment]
> > [UnnecessaryModifier](https://pmd.github.io/pmd/pmd_rules_java_codestyle.html#unnecessarymodifier) recommends to remove public static on a nested interface defined within a class.
> 
> I think this is the real false positive though. The `public` is not unnecessary. If you don't remove the public then CommentDefaultAccessModifier should be quiet.

@harbulot  Are we sure the rule is flagging the `public` modifier and not only the `static` one?

**Comment by harbulot** [comment]
@jsotuyod Good point, I've re-run a test with this code (added `public interface TestInterfaceD`), but nothing shows against `TestInterfaceD`:

```java
public class Example {
    interface TestInterfaceA { // Flags CommentDefaultAccessModifier
        void noop();
    }
    
    static interface TestInterfaceB { // Flags CommentDefaultAccessModifier and UnnecessaryModifier
        void noop();
    }
    
    public static interface TestInterfaceC { // Flags UnnecessaryModifier
        void noop();
    }
    
    public interface TestInterfaceD {
        void noop();
    }
}
```

**Comment by jsotuyod** [comment]
@harbulot awesome! In that case, it seems as both rules are performing as expected. I'm closing this issue.

We may open a new issue on improving `UnnecessaryModifier` to include the redundant modifiers in the message if you please.

**Comment by oowekyala** [comment]
I don't understand how the original issue showed up then. @harbulot Is it possible that when you wrote 

> [UnnecessaryModifier](https://pmd.github.io/pmd/pmd_rules_java_codestyle.html#unnecessarymodifier) recommends to remove `public static` on a nested interface defined within a class.

the rule in fact only recommended to remove `static`?

With a `public static` interface nested in a class, I see the rule only reports the following message: `Unnecessary modifier 'static' on interface 'Bar': member interfaces are implicitly static`

**Comment by harbulot** [comment]
@oowekyala Really sorry, it's probably because I misread the full message when it came up against a `public static interface`:

> Fields in interfaces and annotations are automatically public static final, and methods are public abstract. Classes, interfaces or annotations nested in an interface or annotation are automatically public static (all nested interfaces and annotations are automatically static). Nested enums are automatically static. For historical reasons, modifiers which are implied by the context are accepted by the compiler, but are superfluous.

(It's a completely unrelated Eclipse PMD plugin bug, but the short message in its "*Show details...*" dialog says "*Unnecessary modifier{0} on {1} ''{2}''{3}*`, without filling the placeholders, so I relied on full generic message, and read it too quickly...)

This talks about "*Classes, interfaces or annotations nested in an **interface** or **annotation***", but not nested in classes.

As @jsotuyod said, it might be worth rephrasing that error message to distinguish cases nested in classes too a bit more explicitly (but I should probably have read it a bit more carefully).
